solr/core/src/java/org/apache/solr/storage/AccessDirectory.java

```diff
@@ -180,11 +180,30 @@ public void rename(String source, String dest) throws IOException {
 
   @Override
   public void sync(Collection<String> names) throws IOException {
-    super.sync(names);
-    // NOTE: just call super here. `sync()` should only be called with an expectation
-    // that all files should be present -- notably, sync'ing a lazy partially-populated
-    // file would be pointless (sync'ing is about guaranteeing that a file on disk is
-    // complete, which by definition a lazily reconstituted file is not).
+    ensureOpen();
+
+    for (String name : names) {
+      try {
+        fsync(name); // this covers the normal, actively-writing case.
+      } catch (NoSuchFileException | FileNotFoundException ex) {
+        try {
+          if (activeLazy.get(name) != null
+              || CompressingDirectory.readLengthFromHeader(compressedPath.resolve(name)) > 0) {
```

Comment:
just remind me : we write to persistent and access file parallely, right?

Reply:
Always write to persistent copy first, then access. Sync is parallel, but failure to sync to either access or persistent causes overall sync failure (so when TeeDirectory returns, it's guaranteed to have sync'd the persistent copy)

Reply:
you mean access copy will be sync'ed eventually?